### PR TITLE
Do not add additionalTrailers to ResponseHeaders

### LINKED ISSUE #2544: Trailers is added to ResponseHeaders when there's no content
Related: https://github.com/line/armeria/pull/2372#discussion_r386025977

I found out that the additional trailers is added to `ResponseHeaders` when the content of the response is empty.
I think we implemented this because of gRPC service that sends trailers only when in certain circumstances. https://github.com/line/armeria/blob/master/grpc/src/main/java/com/linecorp/armeria/server/grpc/ArmeriaServerCall.java#L323

However, I think a user may want to send the response with just `ResponseHeaders` and trailers because it does not violate the [spec](https://tools.ietf.org/html/rfc7540#section-8.1).

Perhaps, we can just add the trailers to `ResponseHeaders` in the gRPC service only and remove the trailers from ctx so that `HttpResponseSubscriber` is able to send `ResponseHeaders` and trailers when it should.

/cc @anuraaga, @jyblue

### PR DESCRIPTION
**Pull request by minwoox**
Motivation:
Currently, we combine `additionalResponseTrailers` to the `ResponseHeaders` if the `EOS` of the `ResponseHeaders` is true.
I think we do this because, in gRPC, there's a chance that only trailers is sent to the client so we combine them.
However, the response can consist of just only `ResponseHeaders` and trailers. (i.e no data)
So I think we can just combine the trailers to `ResponseHeaders` in gRPC especially, and let the user send the trailers without data.

Modifications:
- Add `mutateAdditionalResponse(Headers|Trailers)` which takes a `Consumer`
- Add `mutateAdditionalRequestHeaders` which takes a `Consumer`
- Throw an exception when `HttpResponse` and `AggregatedHttpResponse` is created with the `ResponseHeaders` whose status is informational.

Result:
- Close #2544
- You now can send the trailers even when the content is empty.
- (Breaking)
  - Methods adding and removing additional headers and trailers are gone from `ClientRequestContext` and `ServiceRequestContext` except:
    - `addAdditionalResponseHeader()`
    - `setAdditionalResponseHeader()`
    - `addAdditionalResponseTrailer()`
    - `setAdditionalResponseTrailer()`
    - `addAdditionalRequestHeader()`
    - `setAdditionalRequestHeader()`

### CODE REVIEW

File: core/src/main/java/com/linecorp/armeria/server/ServiceRequestContext.java [hunk 1]
```diff
@@ -564,19 +565,21 @@ default String pathParam(String name) {
...
     HttpHeaders additionalResponseTrailers();
 
+    /**
+     * Returns the {@link HttpHeaders} which is set using {@code (add|set)AdditionalResponseTrailer} and
+     * clears the {@link HttpHeaders} so that the {@link HttpHeaders} is not sending twice when
+     * a {@link Service} completes an {@link HttpResponse}.
+     *
+     * <p>This is used only when the trailers is sent as a {@link ResponseHeaders} in gRPC.
+     */
+    HttpHeaders getAndRemoveAdditionalResponseTrailers();
```
Comment (anuraaga) [inline core/src/main/java/com/linecorp/armeria/server/ServiceRequestContext.java]: I don't think we should add a gRPC only API to request context interface. Actually I was expecting a gRPC only implementation to look a little hacky hence encouraging a client option or something similarly at the client level instead :)

Comment (anuraaga) [reply core/src/main/java/com/linecorp/armeria/server/ServiceRequestContext.java]: That being said if we can expose this through an internal API that relies on downcast to `DefaultRequestContext` it's still ok

Comment (minwoox) [reply core/src/main/java/com/linecorp/armeria/server/ServiceRequestContext.java]: Ah, Thanks. Yes, I was a little worried about it but just didn't make sure. 😉 Let me remove this.

File: grpc/src/main/java/com/linecorp/armeria/server/grpc/ArmeriaServerCall.java [hunk 2]
```diff
@@ -534,6 +531,12 @@ static HttpHeaders statusToTrailers(
                          Base64.getEncoder().encodeToString(proto.toByteArray()));
         }
 
+        if (ctx instanceof DefaultServiceRequestContext) {
```
Comment (anuraaga) [inline grpc/src/main/java/com/linecorp/armeria/server/grpc/ArmeriaServerCall.java]: I forget if request context is unwrappable but if so better to use unwrap.

Comment (minwoox) [reply grpc/src/main/java/com/linecorp/armeria/server/grpc/ArmeriaServerCall.java]: It's not unwrappable so I guess just using `instanceof` is the best I can do. 😆

File: core/src/main/java/com/linecorp/armeria/common/AggregatedHttpResponse.java [hunk 3]
```diff
@@ -40,6 +41,9 @@
      * Creates a new HTTP response.
      *
      * @param statusCode the HTTP status code
+     *
+     * @throws IllegalArgumentException if the specified {@code statusCode} is
+     *                                  {@link HttpStatus#isInformational()}.
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/common/AggregatedHttpResponse.java]: How about using `{@linkplain}`?

```suggestion
     * @throws IllegalArgumentException if the specified {@code statusCode} is
     *                                  {@linkplain HttpStatus#isInformational() informational}.
```

File: core/src/main/java/com/linecorp/armeria/common/AggregatedHttpResponse.java [hunk 4]
```diff
@@ -49,9 +53,13 @@ static AggregatedHttpResponse of(int statusCode) {
      * Creates a new HTTP response.
      *
      * @param status the HTTP status
+     *
+     * @throws IllegalArgumentException if the specified {@link HttpStatus} is
+     *                                  {@link HttpStatus#isInformational()}.
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/common/AggregatedHttpResponse.java]:
```suggestion
     * @throws IllegalArgumentException if the specified {@link HttpStatus} is
     *                                  {@linkplain HttpStatus#isInformational() informational}.
```

File: core/src/main/java/com/linecorp/armeria/common/AggregatedHttpResponse.java [hunk 5]
```diff
@@ -159,6 +185,9 @@ static AggregatedHttpResponse of(HttpStatus status, MediaType mediaType, HttpDat
      * Creates a new HTTP response with empty content.
      *
      * @param headers the HTTP headers
+     *
+     * @throws IllegalArgumentException if the status of the specified {@link ResponseHeaders} is
+     *                                  {@link HttpStatus#isInformational()}.
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/common/AggregatedHttpResponse.java]:
```suggestion
     * @throws IllegalArgumentException if the status of the specified {@link ResponseHeaders} is
     *                                  {@linkplain HttpStatus#isInformational() informational}.
```

File: core/src/main/java/com/linecorp/armeria/common/AggregatedHttpResponse.java [hunk 6]
```diff
@@ -198,12 +233,18 @@ static AggregatedHttpResponse of(ResponseHeaders headers, HttpData content, Http
...
      */
     static AggregatedHttpResponse of(Iterable<ResponseHeaders> informationals, ResponseHeaders headers,
                                      HttpData content, HttpHeaders trailers) {
 
         requireNonNull(informationals, "informationals");
         requireNonNull(headers, "headers");
+        final HttpStatus status = headers.status();
+        checkArgument(!status.isInformational(), "status: %s (expected: a non-1xx status", status);
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/common/AggregatedHttpResponse.java]: Missing closing parenthesis:
```suggestion
        checkArgument(!status.isInformational(), "status: %s (expected: a non-1xx status)", status);
```

File: core/src/main/java/com/linecorp/armeria/common/HttpStatus.java [hunk 7]
```diff
@@ -377,14 +377,10 @@ public static HttpStatus valueOf(String statusText) {
...
      * Returns {@code true} if the content of the response for the specified status code is expected to
-     * be always empty (1xx, 204, 205 and 304 responses.)
+     * be always empty (204, 205 and 304 responses.)
      */
     @SuppressWarnings("checkstyle:OverloadMethodsDeclarationOrder")
     public static boolean isContentAlwaysEmpty(int statusCode) {
-        if (HttpStatusClass.INFORMATIONAL.contains(statusCode)) {
-            return true;
-        }
-
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/common/HttpStatus.java]: Why was this removed?

File: core/src/main/java/com/linecorp/armeria/internal/server/annotation/AnnotatedServiceFactory.java [hunk 8]
```diff
@@ -279,12 +279,6 @@ private static HttpStatus defaultResponseStatus(Method method) {
         setAdditionalHeader(defaultTrailers, method, "trailer", methodAlias, "method",
                             AdditionalTrailer.class, AdditionalTrailer::name, AdditionalTrailer::value);
 
-        if (defaultHeaders.status().isContentAlwaysEmpty() && !defaultTrailers.isEmpty()) {
-            logger.warn("A response with HTTP status code '{}' cannot have a content. " +
-                        "Trailers defined at '{}' might be ignored.",
-                        defaultHeaders.status().code(), methodAlias);
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/internal/server/annotation/AnnotatedServiceFactory.java]: Shouldn't we warn because it's true in HTTP/1?

File: core/src/main/java/com/linecorp/armeria/server/DefaultServiceRequestContext.java [hunk 9]
```diff
@@ -496,33 +498,33 @@ public HttpHeaders additionalResponseHeaders() {
...
     public void addAdditionalResponseHeaders(Iterable<? extends Entry<? extends CharSequence, ?>> headers) {
         requireNonNull(headers, "headers");
-        updateAdditionalResponseHeaders(additionalResponseHeadersUpdater,
-                                        builder -> builder.addObject(headers));
+        updateAdditionalHeaders(additionalResponseHeadersUpdater,
+                                builder -> builder.addObject(headers));
     }
 
-    private void updateAdditionalResponseHeaders(
+    private void updateAdditionalHeaders(
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/server/DefaultServiceRequestContext.java]: Not sure why we need to rename this method. Could you explain?

File: core/src/main/java/com/linecorp/armeria/server/DefaultServiceRequestContext.java [hunk 10]
```diff
@@ -534,66 +536,50 @@ private void updateAdditionalResponseHeaders(
         }
     }
 
-    @Override
-    public boolean removeAdditionalResponseHeader(CharSequence name) {
-        return removeAdditionalResponseHeader(additionalResponseHeadersUpdater, name);
-    }
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/server/DefaultServiceRequestContext.java]: I don't get why this method has to be removed. Doesn't a user sometimes need to remove a header from the additional response headers? e.g. A user can write a decorator that removes an additional response header added by other service.

File: core/src/main/java/com/linecorp/armeria/server/DefaultServiceRequestContext.java [hunk 11]
```diff
@@ -534,66 +536,50 @@ private void updateAdditionalResponseHeaders(
...
+    /**
+     * Returns the {@link HttpHeaders} which is set using {@code (add|set)AdditionalResponseTrailer} and
+     * clears the {@link HttpHeaders} so that the {@link HttpHeaders} is not sending twice when
+     * a {@link Service} completes an {@link HttpResponse}.
+     *
+     * <p>This is used only when the trailers is sent as a {@link ResponseHeaders} in gRPC.
+     */
+    public HttpHeaders getAndRemoveAdditionalResponseTrailers() {
+        return additionalResponseTrailersUpdater.getAndSet(this, HttpHeaders.of());
+    }
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/server/DefaultServiceRequestContext.java]: We will not need this method at all if we have `mutateAdditionalResponseTrailers()`:

```java
HttpHeaders oldTrailers = ctx.additionalResponseTrailers();
ctx.mutateAdditionalResponseTrailers(mutator -> mutator.clear());
```

File: core/src/main/java/com/linecorp/armeria/common/HttpStatus.java (hunk 7 above)
Comment (minwoox) [reply core/src/main/java/com/linecorp/armeria/common/HttpStatus.java]: Because I think whether the content is empty or not is decided by the response headers, not the informational headers.
Informational headers will have the following response headers, so shouldn't we look at the response headers to make a decision if the following content will be empty or not?

File: core/src/main/java/com/linecorp/armeria/server/DefaultServiceRequestContext.java (hunk 9 above)
Comment (minwoox) [reply core/src/main/java/com/linecorp/armeria/server/DefaultServiceRequestContext.java]: Let me revert this. Bad decision. 😆

File: core/src/main/java/com/linecorp/armeria/server/DefaultServiceRequestContext.java (hunk 10 above)
Comment (minwoox) [reply core/src/main/java/com/linecorp/armeria/server/DefaultServiceRequestContext.java]: I thought the situation is never going to happen but I guess I was wrong. 😆
And I really like the idea of providing the API to mutate headers. So we can have this behavior again. Thanks!

File: core/src/main/java/com/linecorp/armeria/internal/server/annotation/AnnotatedServiceFactory.java (hunk 8 above)
Comment (minwoox) [reply core/src/main/java/com/linecorp/armeria/internal/server/annotation/AnnotatedServiceFactory.java]: That's a good point. Let me put it back with a little change.

**Comment by minwoox** [comment]
> Perhaps we could replace {add,set,remove}Additional{Request,Response}{Header,Trailer}*() with mutateAdditional{Request,Response}{Headers,Trailers}()?

Fixed. I really like the idea of providing this simple API instead of a bunch of APIs. PTAL. 🙇

File: core/src/main/java/com/linecorp/armeria/client/Clients.java [hunk 12]
```diff
@@ -477,7 +477,7 @@ public static SafeCloseable withHttpHeaders(
         requireNonNull(headerManipulator, "headerManipulator");
         return withContextCustomizer(ctx -> {
             final HttpHeaders manipulatedHeaders = headerManipulator.apply(ctx.additionalRequestHeaders());
-            ctx.setAdditionalRequestHeaders(manipulatedHeaders);
+            ctx.mutateAdditionalRequestHeaders(mutator -> mutator.add(manipulatedHeaders));
```
Comment (minwoox) [inline core/src/main/java/com/linecorp/armeria/client/Clients.java]: In this class, we have 3 different `withHttpHeader` methods and I'm wondering if we can do the same approach here. Or at least we can deprecate this method and add a new method that takes a `Consumer`.

File: core/src/main/java/com/linecorp/armeria/client/ClientRequestContext.java [hunk 13]
```diff
@@ -632,34 +634,9 @@ ClientRequestContext newDerivedContext(RequestId id, @Nullable HttpRequest req,
     HttpHeaders additionalRequestHeaders();
 
     /**
-     * Sets a header with the specified {@code name} and {@code value}. This will remove all previous values
-     * associated with the specified {@code name}.
-     * The header will be included when a {@link Client} sends an {@link HttpRequest}.
-     */
-    void setAdditionalRequestHeader(CharSequence name, Object value);
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/client/ClientRequestContext.java]: How about reviving two methods - `{add,set}AdditionalRequestHeader(CharSequence, Object)` - since it's used pretty commonly?

File: core/src/main/java/com/linecorp/armeria/client/Clients.java (hunk 12 above)
Comment (trustin) [reply core/src/main/java/com/linecorp/armeria/client/Clients.java]: It's pretty common to specify a single header, so I'd prefer keeping them all.

File: core/src/main/java/com/linecorp/armeria/common/HttpStatus.java (hunk 7 above)
Comment (trustin) [reply core/src/main/java/com/linecorp/armeria/common/HttpStatus.java]: Doesn't it still mean that informational headers have no content?

File: core/src/main/java/com/linecorp/armeria/client/Clients.java (hunk 12 above)
Comment (minwoox) [reply core/src/main/java/com/linecorp/armeria/client/Clients.java]: Then just how about changing this method?
```java
// before
public static SafeCloseable withHttpHeaders(
            Function<? super HttpHeaders, ? extends HttpHeaders> headerManipulator) {
// after
public static SafeCloseable withHttpHeaders(Consumer<HttpHeadersBuilder> headerManipulator) {
```
So that a user does not have to `toBuilder` and `build`?

File: core/src/main/java/com/linecorp/armeria/client/ClientRequestContext.java (hunk 13 above)
Comment (minwoox) [reply core/src/main/java/com/linecorp/armeria/client/ClientRequestContext.java]: Yes, let me put this back. 😉

File: core/src/main/java/com/linecorp/armeria/client/Clients.java (hunk 12 above)
Comment (trustin) [reply core/src/main/java/com/linecorp/armeria/client/Clients.java]: OK!

File: core/src/main/java/com/linecorp/armeria/common/HttpStatus.java (hunk 7 above)
Comment (trustin) [reply core/src/main/java/com/linecorp/armeria/common/HttpStatus.java]: Had a quick chat with @minwoox and we chose to keep the modification in this PR, because we can't tell if the response will have no content from an informational status.

File: core/src/main/java/com/linecorp/armeria/client/Clients.java (hunk 12 above)
Comment (minwoox) [reply core/src/main/java/com/linecorp/armeria/client/Clients.java]: Done! 😉

**Comment by ikhoon** [comment]
Still LGTM :-)

**Comment by trustin** [comment]
Thanks a lot, @minwoox !

**Comment by minwoox** [comment]
Thanks for reviewing! 😉
